qiskit/qpy/__init__.py

```diff
@@ -398,6 +425,141 @@ def open(*args):
+interpretation of one field of the ``INSTRUCTION`` struct in an ABI-compatible manner, and adds a
+``INSTRUCTION_ANNOTATIONS`` trailer to ``INSTRUCTION`` which is present conditional on a set bit in
+the ``INSTRUCTION`` payload.
+
+New ANNOTATION_HEADER
+~~~~~~~~~~~~~~~~~~~~~
+
+The ``ANNOTATION_HEADER`` field is a variable-size payload in the header.  It begins with an
+instance of ``ANNOTATION_HEADER_STATIC``, which is the C struct:
+
+.. code-block:: c
+
+    struct ANNOTATION_HEADER_STATIC {
+        uint32_t num_namespaces;
+    }
+
+This is immediately followed by ``num_namespaces`` instances of the ``ANNOTATION_STATE`` payload.
+The order of these is important and should be retained during the deserialization process, as
+subsequent ``INSTRUCTION_ANNOTATION`` payloads will index into it.
+
+The ``ANNOTATION_STATE`` payload begins with the fixed C struct:
+
+.. code-block:: c
+
+    struct ANNOTATION_STATE_HEADER {
+        uint32_t namespace_size;
+        uint64_t state_size;
+    }
+
+This header is immediately followed by ``namespace_size`` bytes of UTF-8 encoded text, which
+comprise the namespace.  Those bytes are immediately followed by ``state_size`` bytes of arbitrary
+data.  The format of this "state" payload is not defined by QPY.  Instead, it is the responsibility
+of an external object associated with the stored namespace.  The format does not dictate how to
+produce these objects; as annotations are entirely custom, the user must supply the serialization
+and deserialization methods.
+
+
+.. _qpy_instruction_v15:
+
+Changes to INSTRUCTION
+~~~~~~~~~~~~~~~~~~~~~~
+
+The ``INSTRUCTION`` struct is modified in an ABI compatible manner to :ref:`its previous definition
+in version 9 <qpy_instruction_v9>`.  The new struct is the C struct (recall that there is no padding
+between any fields, nor at the end of the struct):
+
+.. code-block:: c
+
+    struct INSTRUCTION {
+        uint16_t name_size;
+        uint16_t label_size;
+        uint16_t num_parameters;
+        uint32_t num_qargs;
+        uint32_t num_cargs;
+        uint8_t extras_key;
+        uint16_t conditional_reg_name_size;
+        int64_t conditional_value;
+        uint32_t num_ctrl_qubits;
+        uint32_t ctrl_state;
+    }
+
+where the field ``uint8_t extras_key`` replaces the previous ``uint8_t conditional_key``.  The
+difference is purely in interpretation.  The low two bits of the byte are still interpreted as
+defining the condition and its type.  The high bit of the byte is now a flag, indicated whether an
+``INSTRUCTION_ANNOTATIONS_HEADER`` field is present (if the bit is set) in the trailing data of the
+``INSTRUCTION`` struct.
+
+A complete instruction payload appears in the data stream, including trailing objects and without
+any padding bytes inbetween elements, as:
+
+.. code-block:: text
+
+    struct INSTRUCTION;
+    uint8_t name[name_size];
+    uint8_t label[label_size];
+    uint8_t register[conditional_reg_name_size]; (1)
```

Comment:
Should this `(1)` be after `struct INSTRUCTION` instead?

Reply:
I'd intended it to be a note read as part of understanding the line it's actually on - I can say from just having _used_ the note myself during debugging that it at least made sense to me haha.  I can see the argument for it being on the `struct INSTRUCTION` line if you prefer.